### DISCUSSION TITLE
TetrahedralFEMForceField

### DISCUSSION THREAD

**Discussion by yas28elg** [sofa-framework/sofa#4233] [Develop in C++]
Hi! 
I'm looking for information about TetrahedralFEMForceField. Which elastic model is used for the mechanical constitutive laws? And when is it better to use this force field instead of TetrahedronHyperelasticityFEMForceField ?

Thanks !

**Comment by bakpaul** [comment]
Hi,

I don't know of any component called `TetrahedralFEMForceField` so I think you are either referring to the `TetrahedralCorotationalFEMForceField` or the `TetrahedronFEMForcefield`. Those two components are based on a corotational formulation of the linear elastic constitutive law (yes both).  The later one is the want you should use because it is more up to date and broadly use, and to my knowledge does more than the first one (provide more features). 

Corotational formulation allows to compute the elements' strain independently from their rotation, avoiding to have an augmentation of the internal stress when big rotation occurs. This allows to keep a good precision for large bending of the model when the strain by element stays low.  

In the `TetrahedronHyperelasticityFEMForceField` there is multiple constitutive laws implemented (ArrudaBoyce StVenantKirchhoff, NeoHookean, MooneyRivlin, Ogden for the most famous ones). But their stability limits are not the same provided by the corotational forcefield. 

In fact to know which one you would like to use, you have two criteria : stability and precision
-  **Stability+ / Precision-** Corotational linear elastic model offers a good stability even with large time step. Allowing for the simulation of large object in real time. But of course because it is linear, it is not that precise in term of stress-strain relation and will generate wrong behavior with large deformation.
-   **Stability- / Precision+** Hyperelastic model on the other hand is able to simulation non linear relationship. But because this relationship is linearized at each timestep around the current deformation to enable linear resolution, it tends to have bad stability when using large time step. So using it, you will need to have very small time step and thus will not always be able to have real-time computation.

**Reply by yas28elg** [reply]
Hi, 

Thanks for the clarification, it's very useful!
